File: the-dynamodb-atomic-counter/csharp/src/TheDynamodbAtomicCounter/TheDynamodbAtomicCounterStack.cs
using Amazon.CDK;
using dynamoDB = Amazon.CDK.AWS.DynamoDB;
using apiGateway = Amazon.CDK.AWS.APIGateway;
using iam = Amazon.CDK.AWS.IAM;
using awsSdkCall = Amazon.CDK.CustomResources.AwsSdkCall;
using awsCustomResource = Amazon.CDK.CustomResources.AwsCustomResource;
using awsCustomResourcePolicy = Amazon.CDK.CustomResources.AwsCustomResourcePolicy;
using physicalResourceId = Amazon.CDK.CustomResources.PhysicalResourceId;
using System.Collections.Generic;
using Amazon.CDK.CustomResources;

namespace TheDynamodbAtomicCounter
{

    public class TheDynamodbAtomicCounterStack : Stack
    {
        readonly private string _tableName = "atomicCounter";
        readonly private string _defaultResource = "counter";
        readonly private iam.Role _iamRoleForApiGateway;
        readonly private iam.Role _iamRoleForLambda;
        readonly private dynamoDB.Attribute _partitionKey;
        readonly private dynamoDB.Table _tableAtomicCounter;
        readonly private apiGateway.RestApi _apiGatewayAtomic;
        readonly private apiGateway.AwsIntegration _awsIntegragion;


        internal TheDynamodbAtomicCounterStack(Construct scope, string id, IStackProps props = null) : base(scope, id, props)
        {
            /*
             * 1: Creating appropriate roles
             */

            // IAM Role for APIGateway
            _iamRoleForApiGateway = new iam.Role(this, "iam-role-apigw-stack-cdk", new iam.RoleProps
            {
                RoleName = "iam-role-apigw-stack-cdk",
                AssumedBy = new iam.ServicePrincipal("apigateway.amazonaws.com"),
                ManagedPolicies = new[] { iam.ManagedPolicy.FromAwsManagedPolicyName("AmazonDynamoDBFullAccess") },
                Path = "/service-role/"
            });

            // IAM Role for Lambda - Data insert
            _iamRoleForLambda = new iam.Role(this, "iam-role-lambda-stack-cdk", new iam.RoleProps
            {
                RoleName = "iam-role-lambda-stack-cdk",
                AssumedBy = new iam.ServicePrincipal("lambda.amazonaws.com"),
                ManagedPolicies = new[] { iam.ManagedPolicy.FromAwsManagedPolicyName("AmazonDynamoDBFullAccess") },
                Path = "/service-role/"
            });

            /*
             * 2: Creating the DynamoDB table with a Partition(PrimaryKey)
             */
            _partitionKey = new dynamoDB.Attribute
            {
                Name = "atomicCounter",
                Type = dynamoDB.AttributeType.STRING
            };

            _tableAtomicCounter = new dynamoDB.Table(this, "dynamodb-table", new dynamoDB.TableProps
            {
                TableName = _tableName,
                PartitionKey = _partitionKey,
                RemovalPolicy = RemovalPolicy.DESTROY
            });

            /*
             * 3: We must insert initial data into DynamoDB table
             */

            var customPolicy = awsCustomResourcePolicy.FromSdkCalls( new SdkCallsPolicyOptions
            {
                Resources = awsCustomResourcePolicy.ANY_RESOURCE
            });

            // Params for initial insert
            Dictionary<string, object> createParams = new Dictionary<string, object>
            {
                { "TableName", _tableName},
                { "Item", new Dictionary<string, object>
                    {
                        { "atomicCounter", new Dictionary<string, string>
                            {
                                { "S", "system-aa" }
                            }
                        },
                        { "counterValue", new Dictionary<string, string>
                            {
                                { "N", "1" }
                            }
                        },
                    }
                },
                { "ConditionExpression",  "attribute_not_exists(atomicCounter)" }
            };

            // AwsSdkCall for putItem operation
            // https://docs.aws.amazon.com/cdk/api/latest/python/aws_cdk.custom_resources/AwsSdkCall.html
            var dataTable = new awsSdkCall
            {
                Service = "DynamoDB",
                Action = "putItem",
                PhysicalResourceId = physicalResourceId.Of("data-table"),
                Parameters = createParams
            };

            // AwsCustomResource for putItem operation
            // https://docs.aws.amazon.com/cdk/api/latest/python/aws_cdk.custom_resources/AwsCustomResource.html
            var dataResource = new awsCustomResource(this, "custom-populate-ddb", new AwsCustomResourceProps
            {
                Policy = customPolicy,
                LogRetention = 0,
                Role = _iamRoleForLambda,
                OnCreate = dataTable,
                OnUpdate = dataTable
            });

            // The CFN must wait to finish the table creation before insert the data
            dataResource.Node.AddDependency(_tableAtomicCounter);

            /*
             * 4: Creating Rest API Gateway
             */

            _apiGatewayAtomic = new apiGateway.RestApi(this, "atomic-counter-api", new apiGateway.RestApiProps
            {
                RestApiName = "Atomic Count API",
                Description = "This API serve an Atomic Count API pattern.",
                EndpointTypes = new[] { apiGateway.EndpointType.REGIONAL },
                Deploy = true
            });

            /*
            * 5: Creating AWS Integration Service Response, Integration Options and Integration Service Method
            */

            Dictionary<string, string> responseTemplate = new Dictionary<string, string>();
            responseTemplate.Add("application/json", "#set($value = $input.json('Attributes.counterValue.N'))\r\n#set($l = $value.length())\r\n#set($l = $l - 1)\r\n$value.substring(1,$l)");

            Dictionary<string, string> requestTemplate = new Dictionary<string, string>();
            requestTemplate.Add("application/json", "{\r\n    \"TableName\": \""+_tableName+ "\",\r\n    \"Key\": {\r\n        \"atomicCounter\": {\r\n            \"S\": \"$input.params('systemKey')\"\r\n        }\r\n    },\r\n    \"UpdateExpression\": \"set counterValue = counterValue + :num\",\r\n    \"ExpressionAttributeValues\": {\r\n        \":num\": {\"N\": \"1\"}\r\n    },\r\n    \"ReturnValues\" : \"UPDATED_OLD\"\r\n}");
            
            _awsIntegragion = new apiGateway.AwsIntegration(new apiGateway.AwsIntegrationProps
            {
                Service = "dynamodb",
                Action = "UpdateItem",
                IntegrationHttpMethod = "POST",
                Options = new apiGateway.IntegrationOptions
                {
                    CredentialsRole = _iamRoleForApiGateway,
                    IntegrationResponses = new[] { new apiGateway.IntegrationResponse
                        {
                            StatusCode = "200",
                            ResponseTemplates = responseTemplate
                        }
                    },
                    PassthroughBehavior = apiGateway.PassthroughBehavior.NEVER,
                    RequestTemplates = requestTemplate
                }
            });

            /*
            * 6: Adding resource /counter and method GET.
            */
            Dictionary<string, apiGateway.IModel> responseModel = new Dictionary<string, apiGateway.IModel>();
            responseModel.Add("application/json", apiGateway.Model.EMPTY_MODEL);

            var methodApiGatewayAtomic = _apiGatewayAtomic.Root.AddResource(_defaultResource);
            methodApiGatewayAtomic.AddMethod("GET", _awsIntegragion, new apiGateway.MethodOptions
            {
                MethodResponses = new[] { new apiGateway.MethodResponse
                    {
                        StatusCode = "200",
                        ResponseModels = responseModel
                    }
                }
            });

            /*
            * 7: Output URL
            * The output url will be: https://xxx.execute-api.REGION.amazonaws.com/prod/counter
            */
            new CfnOutput(this, "apigw-counter-url", new CfnOutputProps
            {
                Value = _apiGatewayAtomic.Url + _defaultResource + "?systemKey=system-aa"
            });
        }
    }
}

File: the-dynamodb-atomic-counter/java/src/main/java/com/cdkpatterns/TheDynamodbAtomicCounterStack.java
package com.cdkpatterns;

import java.util.List;
import java.util.Map;

import software.amazon.awscdk.core.CfnOutput;
import software.amazon.awscdk.core.Construct;
import software.amazon.awscdk.core.RemovalPolicy;
import software.amazon.awscdk.core.Stack;
import software.amazon.awscdk.core.StackProps;
import software.amazon.awscdk.services.iam.Role;
import software.amazon.awscdk.services.iam.ServicePrincipal;
import software.amazon.awscdk.services.iam.ManagedPolicy;
import software.amazon.awscdk.services.dynamodb.Table;
import software.amazon.awscdk.services.dynamodb.Attribute;
import software.amazon.awscdk.services.dynamodb.AttributeType;
import software.amazon.awscdk.customresources.AwsCustomResource;
import software.amazon.awscdk.customresources.AwsCustomResourcePolicy;
import software.amazon.awscdk.customresources.AwsCustomResourceProps;
import software.amazon.awscdk.customresources.AwsSdkCall;
import software.amazon.awscdk.customresources.PhysicalResourceId;
import software.amazon.awscdk.customresources.SdkCallsPolicyOptions;
import software.amazon.awscdk.services.apigateway.RestApi;
import software.amazon.awscdk.services.apigateway.EndpointType;
import software.amazon.awscdk.services.apigateway.IntegrationOptions;
import software.amazon.awscdk.services.apigateway.IntegrationResponse;
import software.amazon.awscdk.services.apigateway.MethodOptions;
import software.amazon.awscdk.services.apigateway.MethodResponse;
import software.amazon.awscdk.services.apigateway.Model;
import software.amazon.awscdk.services.apigateway.PassthroughBehavior;
import software.amazon.awscdk.services.apigateway.Resource;
import software.amazon.awscdk.services.apigateway.AwsIntegration;


public class TheDynamodbAtomicCounterStack extends Stack {
	
	private String tableName = "atomicCounter";
    private String defaultResource = "counter";
	
    public TheDynamodbAtomicCounterStack(final Construct scope, final String id) {
        this(scope, id, null);
    }

    public TheDynamodbAtomicCounterStack(final Construct scope, final String id, final StackProps props) {
        super(scope, id, props);

        /*
         * 1: Creating appropriate roles
         */

        // IAM Role for APIGateway
        Role iamRoleForApiGateway = Role.Builder.create(this, "iam-role-apigw-stack-cdk")
        		.roleName("iam-role-apigw-stack-cdk")
        		.assumedBy(new ServicePrincipal("apigateway.amazonaws.com"))
        		.managedPolicies(List.of(
        				ManagedPolicy.fromAwsManagedPolicyName("AmazonDynamoDBFullAccess")
        				))
        		.path("/service-role/")
        		.build();

        // IAM Role for Lambda - Data insert
        Role iamRoleForLambda = Role.Builder.create(this, "iam-role-lambda-stack-cdk")
        		.roleName("iam-role-lambda-stack-cdk")
        		.assumedBy(new ServicePrincipal("lambda.amazonaws.com"))
        		.managedPolicies(List.of(
        				ManagedPolicy.fromAwsManagedPolicyName("AmazonDynamoDBFullAccess")
        				))
        		.path("/service-role/")
        		.build();
        
        /*
         * 2: Creating the DynamoDB table with a Partition(PrimaryKey)
         */
        Attribute partitionKey = new Attribute
        		.Builder()
        		.name("atomicCounter")
        		.type(AttributeType.STRING)
        		.build();

        Table tableAtomicCounter = Table.Builder.create(this, "dynamodb-table")
        		.tableName(tableName)
        		.partitionKey(partitionKey)
        		.removalPolicy(RemovalPolicy.DESTROY)
        		.build();
        
        /*
         * 3: We must insert initial data into DynamoDB table
         */
        AwsCustomResourcePolicy customPolicy = AwsCustomResourcePolicy.fromSdkCalls(
        		new SdkCallsPolicyOptions.Builder().resources(AwsCustomResourcePolicy.ANY_RESOURCE)
        		.build()
        		);
        
        // Params for initial insert
        Map<String, Object> createParams = Map.ofEntries(
        		Map.entry("TableName", tableName),
                Map.entry("Item", Map.ofEntries(
                        Map.entry("atomicCounter", Map.of("S", "system-aa")),
                        Map.entry("counterValue", Map.of("N", "1"))
                )),
                Map.entry("ConditionExpression", "attribute_not_exists(atomicCounter)"));
        
        // AwsSdkCall for putItem operation
        // https://docs.aws.amazon.com/cdk/api/latest/python/aws_cdk.custom_resources/AwsSdkCall.html
        AwsSdkCall dataTable = new AwsSdkCall.Builder()
        		.service("DynamoDB")
        		.action("putItem")
        		.physicalResourceId(PhysicalResourceId.of("data-table"))
        		.parameters(createParams)
        		.build();
        
        // AwsCustomResource for putItem operation
        // https://docs.aws.amazon.com/cdk/api/latest/python/aws_cdk.custom_resources/AwsCustomResource.html
        AwsCustomResource dataResource = new AwsCustomResource(this, "custom-populate-ddb",
        		new AwsCustomResourceProps.Builder()
        		.policy(customPolicy)
        		.logRetention(null)
        		.role(iamRoleForLambda)
        		.onCreate(dataTable)
        		.onUpdate(dataTable)
        		.build());
        
        // The CFN must wait to finish the table creation before insert the data
        dataResource.getNode().addDependency(tableAtomicCounter);

        /*
         * 4: Creating Rest API Gateway
         */
        RestApi apiGatewayAtomic = RestApi.Builder.create(this, "atomic-counter-api")
        		.restApiName("Atomic Count API")
        		.description("This API serve an Atomic Count API pattern.")
        		.endpointTypes(List.of(EndpointType.REGIONAL))
        		.deploy(true)
        		.build();
        
        
        String responseTemplate = "#set($value = $input.json('Attributes.counterValue.N'))\r\n#set($l = $value.length())\r\n#set($l = $l - 1)\r\n$value.substring(1,$l)";
        String requestTemplate = "{\r\n    \"TableName\": \""+tableName+ "\",\r\n    \"Key\": {\r\n        \"atomicCounter\": {\r\n            \"S\": \"$input.params('systemKey')\"\r\n        }\r\n    },\r\n    \"UpdateExpression\": \"set counterValue = counterValue + :num\",\r\n    \"ExpressionAttributeValues\": {\r\n        \":num\": {\"N\": \"1\"}\r\n    },\r\n    \"ReturnValues\" : \"UPDATED_OLD\"\r\n}";
        
        AwsIntegration awsIntegragion = AwsIntegration
        		.Builder
        		.create()
        		.service("dynamodb")
        		.action("UpdateItem")
        		.integrationHttpMethod("POST")
        		.options(new IntegrationOptions.Builder()
        				.credentialsRole(iamRoleForApiGateway)
        				.integrationResponses(List.of(
        						new IntegrationResponse
        						.Builder()
        						.statusCode("200")
        						.responseTemplates(Map.of(
        								"application/json", responseTemplate))
        						.build()
        						))
        				.passthroughBehavior(PassthroughBehavior.NEVER)
        				.requestTemplates(Map.of(
        						"application/json", requestTemplate))
        				.build())
        		.build();
        
        		
        /*
         * 6: Adding resource /counter and method GET.
         */
        Resource methodApiGatewayAtomic = apiGatewayAtomic.getRoot().addResource(defaultResource);
        methodApiGatewayAtomic.addMethod("GET", awsIntegragion, new MethodOptions
        		.Builder()
        		.methodResponses(List.of(
        				new MethodResponse
        				.Builder()
        				.statusCode("200")
        				.responseModels(Map.of(
        						"application/json", Model.EMPTY_MODEL
        						))
        				.build()))
        		.build());
        
        /*
         * 7: Output URL
         * The output url will be: https://xxx.execute-api.REGION.amazonaws.com/prod/counter
         */
        CfnOutput.Builder.create(this, "apigw-counter-url")
        	.value(apiGatewayAtomic.getUrl() + defaultResource + "?systemKey=system-aa")
        	.build();
    }
}
